File: Javascript/1.two-sum.js
/*
 * @lc app=leetcode id=1 lang=javascript
 *
 * [1] Two Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    const index = new Map();
    for (let i = 0; i < nums.length; i++) {
        const subpart = target - nums[i];
        index.set(subpart, i);
    }
    for (let i = 0; i < nums.length; i++) {
        const key = index.get(nums[i]);

        if (i === key) {
            continue;
        }

        if (key !== undefined) {
            return [i, key];
        }
    }
};
// @lc code=end

File: Ruby/1.two-sum.rb
#
# @lc app=leetcode id=1 lang=ruby
#
# [1] Two Sum
#

# @lc code=start
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    hash = {}
    for i in 0...nums.size
        hash[target - nums[i]] = i
    end

    for i in 0...nums.size
        if !hash.key? nums[i]
            next
        end
        extracted = hash[nums[i]]
        if extracted == i
            next
        end
        return [i, extracted]
    end
end
# @lc code=end

print two_sum([3,2,4], 6)
